torchok/models/heads/detection/fcos.py

```diff
@@ -0,0 +1,218 @@
+            [points.repeat(num_imgs, 1) for points in all_level_points])
+
+        # FG cat_id: [0, num_classes -1], BG cat_id: num_classes
+        bg_class_ind = self.num_classes
+        pos_inds = ((flatten_labels >= 0) & (flatten_labels < bg_class_ind)).nonzero().reshape(-1)
+        num_pos = torch.tensor(len(pos_inds), dtype=torch.float, device=bbox_preds[0].device)
+        num_pos = max(reduce_mean(num_pos), 1.0)
+
+        pos_bbox_preds = flatten_bbox_preds[pos_inds]
+        pos_centerness = flatten_centerness[pos_inds]
+        pos_bbox_targets = flatten_bbox_targets[pos_inds]
+        pos_centerness_targets = self.centerness_target(pos_bbox_targets)
+        # centerness weighted iou loss
+        centerness_denorm = max(reduce_mean(pos_centerness_targets.sum().detach()), 1e-6)
+
+        # if len(pos_inds) > 0:
+        pos_points = flatten_points[pos_inds]
+        pos_decoded_bbox_preds = self.bbox_coder.decode(pos_points, pos_bbox_preds)
+        pos_decoded_target_preds = self.bbox_coder.decode(pos_points, pos_bbox_targets)
+
+        return joint_loss(
+            flatten_cls_scores=flatten_cls_scores.float(),
+            flatten_labels=flatten_labels,
+            num_pos=num_pos,
+            pos_decoded_bbox_preds=pos_decoded_bbox_preds.float(),
+            pos_decoded_target_preds=pos_decoded_target_preds.float(),
+            pos_centerness_targets=pos_centerness_targets.float(),
+            centerness_denorm=centerness_denorm,
+            pos_centerness=pos_centerness.float()
+        )
+
+    @force_fp32(apply_to=('cls_scores', 'bbox_preds'))
+    def get_bboxes(self, cls_scores, bbox_preds, image_shape, **kwargs):
+        """Transform network outputs of a batch into bbox results.
+
+        Note: When score_factors is not None, the cls_scores are
+        usually multiplied by it then obtain the real score used in NMS,
+        such as CenterNess in FCOS, IoU branch in ATSS.
+
+        Args:
+            cls_scores (list[Tensor]): Classification scores for all
+                scale levels, each is a 4D-tensor, has shape
+                (batch_size, num_priors * num_classes, H, W).
+            bbox_preds (list[Tensor]): Box energies / deltas for all
+                scale levels, each is a 4D-tensor, has shape
+                (batch_size, num_priors * 4, H, W).
+            image_shape (Tuple[int, int]): size of the input image.
+
+        Returns:
+            list[list[Tensor, Tensor]]: Each item in result_list is 2-tuple.
+                The first item is an (n, 5) tensor, where the first 4 columns
+                are bounding box positions (tl_x, tl_y, br_x, br_y) and the
+                5-th column is a score between 0 and 1. The second item is a
+                (n,) tensor where each item is the predicted class label of
+                the corresponding box.
+        """
+        pseudo_meta = [dict(img_shape=image_shape, scale_factor=1) for i in range(len(cls_scores[0]))]
```

Comment:
May you explain me why `get_bboxes` work in mmdet fcos? because default parameter for `img_metas` is None and they have a code line 
```for img_id in range(len(img_metas)):``` so it must fall.

Reply:
In mmdet this function cannot work without `img_metas`. It is a trick that the default value is None. But inside their code the meta is used only to get image shape. Therefore in my implementation I initialize it with base image shape.